### ISSUE TITLE
"error: expression must be a modifiable lvalue"

### ISSUE THREAD

**Issue by yongqli** [AccelerateHS/accelerate#264]
I can't negate twice, as it would get turned into `--`, which is interpreted as C++'s pre-decrement operator rather than `- -`.

In GHCI:

```
import qualified Data.Array.Accelerate                as A
import qualified Data.Array.Accelerate.CUDA           as GPGPU

let x = A.use $ A.fromList (A.Z A.:. 1) [3::Double] :: A.Acc (A.Vector Double)
GPGPU.run $ A.map negate $ A.map negate x
```

**Comment by yongqli** [comment]
Can you make sure that a similar problem isn't happening with other operators, such as `<*` and `>*`? I have a case where I strongly suspect accelerate is generating incorrect code, but it's too time consuming to track down.

**Comment by yongqli** [comment]
Also, how can I look at the AST and C++ code that is generated?

**Comment by tmcdonell** [comment]
This appears to be a problem in `language-c-quote` or its pretty printer, and I submitted an issue there: mainland/language-c-quote#48.

**Comment by tmcdonell** [comment]
seems like this happens with `not` as well, but `gcc` appears happy to compile `!!x`, so I don't think that should be an issue.

**Comment by tmcdonell** [comment]
The usual command line flag should work: `-ddump-cc`.

Note that you'll also have to install the base accelerate package in debug mode for this to work, which has changed. Let me know this is not working for you though.

Also new is that you can set the environment variable `ACCELERATE_FLAGS` to the usual strings you would have typed on the command line, to avoid typing it out every time.

**Comment by yongqli** [comment]
Could `!` somehow get attached to a different expression? Meaning if I have `not (x && y)` end up with `not x && y` somehow? I just tried a simple example and it seems to have worked, but I'm seeing weird things in a loop, where the loop terminates without both of its stopping condition being met.

**Comment by tmcdonell** [comment]
At a small test `not (x && y)` produced the correct code, but perhaps this falls over with some larger example or corner case. If you want me to look at this further, you'll need to provide a test case.

Have you verified that you get the correct results in the interpreter?

**Comment by tmcdonell** [comment]
Fixed via mainland/language-c-quote#48
